// @relaypro/sdk/dist/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimeoutType = exports.TimerType = exports.NotificationSound = exports.NotificationPriority = exports.IncidentStatus = exports.Notification = exports.DeviceType = exports.DeviceInfoField = exports.DeviceInfoQuery = exports.Language = exports.Taps = exports.Button = exports.CallDirection = exports.Event = void 0;
var Event;
(function (Event) {
    Event["ERROR"] = "error";
    Event["START"] = "start";
    Event["STOP"] = "stop";
    Event["BUTTON"] = "button";
    Event["TIMER"] = "timer";
    Event["NOTIFICATION"] = "notification";
    Event["INCIDENT"] = "incident";
    Event["PROMPT_START"] = "prompt_start";
    Event["CALL_RINGING"] = "call_ringing";
    Event["CALL_CONNECTED"] = "call_connected";
    Event["CALL_DISCONNECTED"] = "call_disconnected";
    Event["CALL_FAILED"] = "call_failed";
    Event["CALL_RECEIVED"] = "call_received";
    Event["CALL_START_REQUEST"] = "call_start_request";
})(Event = exports.Event || (exports.Event = {}));
var CallDirection;
(function (CallDirection) {
    CallDirection["INBOUND"] = "inbound";
    CallDirection["OUTBOUND"] = "outbound";
})(CallDirection = exports.CallDirection || (exports.CallDirection = {}));
var Button;
(function (Button) {
    Button["ACTION"] = "action";
    Button["CHANNEL"] = "channel";
})(Button = exports.Button || (exports.Button = {}));
var Taps;
(function (Taps) {
    Taps["SINGLE"] = "single";
    Taps["DOUBLE"] = "double";
    Taps["TRIPLE"] = "triple";
})(Taps = exports.Taps || (exports.Taps = {}));
var Language;
(function (Language) {
    Language["ENGLISH"] = "en-US";
    Language["GERMAN"] = "de-DE";
    Language["SPANISH"] = "es-ES";
    Language["FRENCH"] = "fr-FR";
    Language["ITALIAN"] = "it-IT";
    Language["RUSSIAN"] = "ru-RU";
    Language["SWEDISH"] = "sv-SE";
    Language["TURKISH"] = "tr-TR";
    Language["HINDI"] = "hi-IN";
    Language["ICELANDIC"] = "is-IS";
    Language["JAPANESE"] = "ja-JP";
    Language["KOREAN"] = "ko-KR";
    Language["POLISH"] = "ko-KR";
    Language["PORTUGUESE"] = "pt-BR";
    Language["NORWEIGN"] = "nb-NO";
    Language["DUTCH"] = "nl-NL";
    Language["CHINESE"] = "zh";
})(Language = exports.Language || (exports.Language = {}));
var DeviceInfoQuery;
(function (DeviceInfoQuery) {
    DeviceInfoQuery["NAME"] = "name";
    DeviceInfoQuery["ID"] = "id";
    DeviceInfoQuery["TYPE"] = "type";
    DeviceInfoQuery["ADDRESS"] = "address";
    DeviceInfoQuery["COORDINATES"] = "latlong";
    DeviceInfoQuery["BATTERY"] = "battery";
    DeviceInfoQuery["INDOOR_LOCATION"] = "indoor_location";
    DeviceInfoQuery["LOCATION"] = "location";
})(DeviceInfoQuery = exports.DeviceInfoQuery || (exports.DeviceInfoQuery = {}));
var DeviceInfoField;
(function (DeviceInfoField) {
    DeviceInfoField["LABEL"] = "label";
    DeviceInfoField["CHANNEL"] = "channel";
})(DeviceInfoField = exports.DeviceInfoField || (exports.DeviceInfoField = {}));
var DeviceType;
(function (DeviceType) {
    DeviceType["RELAY"] = "relay";
    DeviceType["RELAY2"] = "relay2";
    DeviceType["RELAY_APP"] = "relay_app";
    DeviceType["ROIP"] = "roip";
    DeviceType["DASH"] = "dash";
})(DeviceType = exports.DeviceType || (exports.DeviceType = {}));
var Notification;
(function (Notification) {
    Notification["BROADCAST"] = "broadcast";
    Notification["ALERT"] = "alert";
    Notification["NOTIFY"] = "notify";
    Notification["CANCEL"] = "cancel";
})(Notification = exports.Notification || (exports.Notification = {}));
var IncidentStatus;
(function (IncidentStatus) {
    IncidentStatus["RESOLVED"] = "resolved";
    IncidentStatus["CANCELLED"] = "cancelled";
})(IncidentStatus = exports.IncidentStatus || (exports.IncidentStatus = {}));
var NotificationPriority;
(function (NotificationPriority) {
    NotificationPriority["NORMAL"] = "normal";
    NotificationPriority["HIGH"] = "high";
    NotificationPriority["CRITICAL"] = "critical";
})(NotificationPriority = exports.NotificationPriority || (exports.NotificationPriority = {}));
var NotificationSound;
(function (NotificationSound) {
    NotificationSound["DEFAULT"] = "default";
    NotificationSound["SOS"] = "sos";
})(NotificationSound = exports.NotificationSound || (exports.NotificationSound = {}));
var TimerType;
(function (TimerType) {
    TimerType["TIMEOUT"] = "timeout";
    TimerType["INTERVAL"] = "interval";
})(TimerType = exports.TimerType || (exports.TimerType = {}));
var TimeoutType;
(function (TimeoutType) {
    TimeoutType["MILLISECONDS"] = "ms";
    TimeoutType["SECONDS"] = "secs";
    TimeoutType["MINUTES"] = "mins";
    TimeoutType["HOURS"] = "hrs";
})(TimeoutType = exports.TimeoutType || (exports.TimeoutType = {}));
